src/Workspaces/Core/Portable/Rename/LightweightRenameLocations.cs

```diff
@@ -0,0 +1,143 @@
+    /// </summary>
+    internal sealed partial class LightweightRenameLocations
+    {
+        public readonly ISymbol Symbol;
+        public readonly Solution Solution;
+        public readonly SymbolRenameOptions Options;
+        public readonly CodeCleanupOptionsProvider FallbackOptions;
+
+        public readonly ImmutableHashSet<RenameLocation> Locations;
+
+        // Note: these two are just stored as arrays so that we can pass them back/forth with a
+        // SerializableRenameLocations object without having to do any work.  As they are never mutated in this type and
+        // are private, this is safe and cheap.
+
+        private readonly SerializableReferenceLocation[]? _implicitLocations;
+        private readonly SerializableSymbolAndProjectId[]? _referencedSymbols;
+
+        private LightweightRenameLocations(
+            ISymbol symbol,
+            Solution solution,
+            SymbolRenameOptions options,
+            CodeCleanupOptionsProvider fallbackOptions,
+            ImmutableHashSet<RenameLocation> locations,
+            SerializableReferenceLocation[]? implicitLocations,
+            SerializableSymbolAndProjectId[]? referencedSymbols)
+        {
+            Symbol = symbol;
+            Solution = solution;
+            Options = options;
+            FallbackOptions = fallbackOptions;
+            Contract.ThrowIfNull(locations);
+            Locations = locations;
+            _implicitLocations = implicitLocations;
+            _referencedSymbols = referencedSymbols;
+        }
+
+        public async Task<HeavyweightRenameLocations?> ToHeavyweightAsync(CancellationToken cancellationToken)
+        {
+            var referencedSymbols = _referencedSymbols is null
+                ? default
+                : await _referencedSymbols.SelectAsArrayAsync(sym => sym.TryRehydrateAsync(Solution, cancellationToken)).ConfigureAwait(false);
+
+            if (!referencedSymbols.IsDefault && referencedSymbols.Any(s => s is null))
+                return null;
+
+            var implicitLocations = _implicitLocations is null
+                ? default
+                : await _implicitLocations.SelectAsArrayAsync(loc => loc.RehydrateAsync(Solution, cancellationToken)).ConfigureAwait(false);
+
+            return new HeavyweightRenameLocations(
+               Symbol,
+               Solution,
+               Options,
+               FallbackOptions,
+               Locations,
+               implicitLocations,
+               referencedSymbols);
+        }
+
+        /// <summary>
+        /// Find the locations that need to be renamed.  Can cross process boundaries efficiently to do this.
+        /// </summary>
+        public static async Task<LightweightRenameLocations> FindRenameLocationsAsync(
+            ISymbol symbol, Solution solution, SymbolRenameOptions options, CodeCleanupOptionsProvider fallbackOptions, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(solution);
```

Comment:
Why are these needed for internal nullable enabled code?

Reply:
I guess this method is just a copy and paste from some old utilities : )

Reply:
yes.  lots of rename is not NRT enabled.  so we're in a mix of patterns here.  it's not great.